src/py/flwr/simulation/ray_transport/ray_actor.py

```diff
@@ -0,0 +1,280 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Ray-based Flower Actor and ActorPool implementation."""
+
+import threading
+import traceback
+from logging import ERROR, WARNING
+from typing import Any, Callable, Dict, List, Set
+
+import ray
+from ray.util.actor_pool import ActorPool
+
+from flwr.common.logger import log
+
+
+class ClientException(Exception):
+    """Raised when client side logic crashes with an exception."""
+
+    def __init__(self, message: str):
+        self.message = f"\n{'>'*7} A ClientException occurred." + message
+        super().__init__(self.message)
+
+
+@ray.remote
+class VirtualClientEngineActor:
+    """A Ray Actor class that runs client workloads."""
+
+    def terminate(self):
+        """Terminate Actor."""
+        log(WARNING, f"Manually terminating {self.__class__.__name__}")
+        ray.actor.exit_actor()
+
+    def run(self, client_fn: Callable, client_id):
+        """Run a client workload."""
+        # execute tasks and return result
+        # return also cid which is needed to ensure results
+        # from the pool are correctly assigned to each ClientProxy
+        try:
+            client_results = client_fn()
+        except Exception as ex:
+            client_trace = traceback.format_exc()
+            message = (
+                "\n\tSomething went wrong when running your client workload."
+                f"\n\tClient {client_id} crashed when the {self.__class__.__name__}"
+                " was running its workload."
+                f"\n\tException triggered on the client side: {client_trace}"
+            )
+            raise ClientException(message) from ex
+
+        return client_id, client_results
+
+
+def pool_size_from_resources(client_resources: Dict):
+    """Calculate number of Actors that fit in pool given the resources in the.
+
+    cluster and those required per client.
+    """
+    cluster_resources = ray.cluster_resources()
+    num_cpus = cluster_resources["CPU"]
+    num_gpus = cluster_resources.get("GPU", 0)  # there might not be GPU
+    num_actors = int(num_cpus / client_resources["num_cpus"])
+    # if a GPU is present and client resources do require one
+    if client_resources["num_gpus"] > 0.0:
+        if num_gpus:
+            # if there are gpus in the cluster
+            num_actors = min(num_actors, int(num_gpus / client_resources["num_gpus"]))
+        else:
+            num_actors = 0
+
+    if num_actors == 0:
+        log(
+            WARNING,
+            f"Your ActorPool is empty. Your system ({num_cpus = }, {num_gpus = }) "
+            "does not meet the criteria to host at least one client with resources:"
+            f" {client_resources}. Consider lowering your `client_resources`",
+        )
```

Comment:
Should this be an error that stops the simulation?

Reply:
ok